docs(seed): Document how to edit and test challenges

alexch opened the pull request:
#### Pre-Submission Checklist
- [x] Your pull request targets the `staging` branch of freeCodeCamp.
- [ ] Branch starts with either `fix/`, `feature/`, or `translate/` (e.g. `fix/signin-issue`)
- [x] You have only one commit (if not, [squash](http://forum.freecodecamp.org/t/how-to-squash-multiple-commits-into-one-with-git/13231) them into one commit).
- [x] All new and existing tests pass the command `npm test`. Use `git commit --amend` to amend any fixes.

#### Type of Change
- [ ] Small bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Add new translation (feature adding new translations)

TODO: add "documentation" to this PR template :-)

#### Checklist:
- [x] Tested changes locally.
- [ ] Addressed currently open issue (replace XXXXX with an issue no in next line)

#### Description
<!-- Describe your changes in detail -->

add seed/README.md documenting scripts and schema for challenges

For discussion. Some of my open questions before merging:

1. Does this doc belong here, or should I merge it into the top level CONTRIBUTING.md file?

2. Did I get the command/field names/descriptions right?

3. I can see  what `node seed` does -- it runs seed/index.js -- but should that code be moved to use `gulp` instead, for conformity with `npm run test` et al.? It's confusing that some scripts are run with `npm run` and others with `node`.

alexch committed:
docs(seed): Document how to edit and test challenges

add seed/README.md documenting scripts and schema for challenges
edit challenge-style-guide with mongodb id generation info

Closes #16373

docs(seed): add Ethan's video, apply PR feedback

alexch commented:
@camperbot it's a documentation PR, so please update your list of acceptable branch name prefixes

raisedadead commented:
Hi @alexch nice work.. thanks for doing this.

alexch commented:
@raisedadead any thoughts on my 3 questions? I'm working on moving my http://testfirst.org/live labs into FCC, and I'm willing to document and clean up code & processes as I go.

raisedadead commented:
@alexch sorry about the delays, in getting back to you. Have got an unusual amount of work these days.
First of all nice work, I'll add some feedback when I get more time to go through this, please bear with me.

Okay, let me clarify your queries:

> Does this doc belong here, or should I merge it into the top level CONTRIBUTING.md file?

This is fine for now, I am actually working on another PR #16432  to document everything in a `docs/` directory. I can move this there keeping a structure, later after merging yours.

Our main contributing guide contains an overload of information, that could be broken into separate parts, for ease of following.

> Did I get the command/field names/descriptions right?

From a quick read, yes they are correct. However, I need a deeper read. Sorry.

> I can see what node seed does -- it runs seed/index.js -- but should that code be moved to use gulp instead, for conformity with npm run test et al.? It's confusing that some scripts are run with npm run and others with node.

Yes, that is correct. And I agree this is a bit confusing, especially to people not familiar with the code-base.

Actually the seeding (...the database) is independent of the actual server, and its business logic, so it was never added to npm or the gulp scripts.

What I mean is we could theoretically rip of the seed directory and noting would ideally break on the app itself.

But this is a confusing aspect, and we should add instructions and unify this into the scripts (although, still keeping them independent of the current scripts).

I guess, you should rather create a fresh issue for this third point.

Other than that, since you are working on this already, would you mind taking on https://github.com/freeCodeCamp/freeCodeCamp/issues/16373 ?

I can take that of my list, as it should be added with your changes I think.

raisedadead commented:
P.S: @camperbot is just a little dumb bot right now, I have yet to update its humor settings.

raisedadead commented:
> I'm working on moving my http://testfirst.org/live labs into FCC, and I'm willing to document and clean up code & processes as I go.

Can you also create a fresh issue for this, we will be happy to discuss about this, on that thread.
Thanks again for your interest.

alexch commented:
No hurry, no worries. You gave a full reply within 24 hours, that's still an A+ rating. :-)

> What I mean is we could theoretically rip of the seed directory and noting would ideally break on the app itself.

Do you think this is a good design goal? Personally I think there's value in keeping them bound together, especially since they share a schema. I like the organization that `client`, `server`, and `seed` are all in one repo, so changes in one part must be reflected in the others, and all tests must pass in concert.

> Other than that, since you are working on this already, would you mind taking on #16373 ?

Sure! Happy to. I'll open issues for my other proposals too.

raisedadead commented:
> Do you think this is a good design goal? 

Yup, that is intentional. The code will always be in this repo, including the curriculum but we want to break a lot of the monolithic architecture.

For example, since the curriculum in seed is independent of server, its possible to update our production database on the fly without shutting down the servers that are getting data from them.

We want to further have the client (react) app to be separate of the server, so that it can be served off CDNs while, we have a more API only backend. This will take us closer towards being offline, async, etc.

There are ton of possibilities that way.

Of course the code stays in this repo, for best developer experience.

alexch commented:
> we want to break a lot of the monolithic architecture... Of course the code stays in this repo, for best developer experience.

👍 and 👍

raisedadead commented on seed/README.md:
```diff
@@ -0,0 +1,61 @@
+# Challenges
+
+The `seed` directory contains all the challenges that appear on the FreeCodeCamp site. 
```
FreeCodeCamp `=>` freeCodeCamp

raisedadead replied:
also, prefer:

`...challenges that appear on the freeCodeCamp learning platform.`

raisedadead commented on seed/README.md:
```diff
@@ -0,0 +1,61 @@
+## Challenge JSON Fields
+
+* [Challenge schema](../common/models/challenge.json) - lists all of the fields inside challenge, and describes some of them
+* [Challenge template](https://gist.github.com/systimotic/e6ab8ba82222b8e0c381ab7dc2d4ff6f) - a gist made by @systimotic, replicated below
```
Unless @systimotic has any other views we can just keep the copy in this readme.

raisedadead commented on seed/challenge-style-guide.md:
````diff
@@ -63,6 +63,39 @@ Here are some example challenge names:
+
+Once you have your id, put it into the JSON file as the `id` field, e.g.
+
+```json
````
` ```json ` `=>` ` ``` `

I know, github is weird.

alexch replied:
my editor (RubyMine) displays here-doc text with styling if you provide it a language hint as above, and automatically inserts it; I'm happy to delete the hint, but I don't think GitHub cares one way or the other, does it?

alexch replied:
![image](https://user-images.githubusercontent.com/8524/35160333-d23431ea-fd0a-11e7-9cef-9c3af7baa7dd.png)
this screengrab shows it's ignoring the `json` after the triple-backtick -- again, I'm happy to delete it if you think it's confusing to future documentarians

alexch replied:
OHHH, it's complaining about the comment in the other file's JSON snippet. Which makes sense since JSON doesn't support comments.

alexch replied:
aha, https://help.github.com/articles/creating-and-highlighting-code-blocks/#syntax-highlighting -- I now think this is a super-cool feature and recommend keeping it

raisedadead reviewed (approved):
Very minor cosmetic changes. We are ready to merge... nice work.

alexch commented:
I'll make those changes, rebase-and-squash, and commit. Also I added a link to Ethan's screencast.

alexch commented on seed/challenge-style-guide.md:
````diff
@@ -63,6 +63,38 @@ Here are some example challenge names:
+{
+  "id": "5a474d78df58bafeb3535d34",
+  "title": "Challenge Title",
+```
````
...and this gets weird again. GitHub's diff algorithm is apparently unaware that the triple-tick closes the JSON block and marking it and the next few lines red. Looking at the entire file shows there is no problem with their actual markdown formatter: 
![image](https://user-images.githubusercontent.com/8524/35163715-aeb19590-fd15-11e7-8bd1-3fb608d350b9.png)

raisedadead replied:
I agree, I think adding the `json` is doing nothing special with regards to highlight other making it unreadable in the future diffs, lets just remove it.

raisedadead replied:
Other than that everything looks good for merge.

raisedadead committed:
docs: fix up syntax highlight

While it was ideal to have ```json, its not really doing anything in terms of highlight. But, the side effect in diff is making the other parts really ugly and unreadable.

Hence fixing up.

raisedadead commented:
#### :sparkles: :sparkles: **LGTM.** :sparkles: :sparkles: 

Thanks @alexch nice work!

Happy Contributing!
